File: android/src/main/kotlin/sncf/connect/tech/flutter_eco_mode/listener/BatteryLevelListener.kt
package sncf.connect.tech.flutter_eco_mode.listener

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import sncf.connect.tech.flutter_eco_mode.BatteryLevelStreamHandler
import sncf.connect.tech.flutter_eco_mode.EcoBatteryManager
import sncf.connect.tech.flutter_eco_mode.MessagesPigeonMethodCodec
import sncf.connect.tech.flutter_eco_mode.PigeonEventSink

class BatteryLevelListener(
    private val ecoBatteryManager: EcoBatteryManager,
) : BatteryLevelStreamHandler(), DisposableStreamListener {
    private var batteryLevelEventSink: PigeonEventSink<Double>? = null
    private var batteryLevelReceiver: BroadcastReceiver? = null
    private var lastSentLevel: Double? = null

    override fun onListen(p0: Any?, sink: PigeonEventSink<Double>) {
        batteryLevelEventSink = sink

        batteryLevelReceiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent?) {
                sendBatteryLevel(intent)
            }
        }

        val stickyIntent = ecoBatteryManager.registerBatteryReceiver(batteryLevelReceiver)
        sendBatteryLevel(stickyIntent)
    }

    override fun onCancel(p0: Any?) {
        cleanUp()
    }

    override fun register(binaryMessenger: BinaryMessenger) = register(binaryMessenger, this)

    override fun dispose(binaryMessenger: BinaryMessenger) {
        cleanUp()

        EventChannel(binaryMessenger, EcoModeEventChannels.BATTERY_LEVEL, MessagesPigeonMethodCodec).setStreamHandler(null)
    }

    private fun cleanUp() {
        batteryLevelReceiver?.let { ecoBatteryManager.unregisterReceiver(it) }
        batteryLevelReceiver = null
        batteryLevelEventSink = null
        lastSentLevel = null
    }

    private fun sendBatteryLevel(intent: Intent?) {
        val batteryPct = ecoBatteryManager.parseLevel(intent)

        if (batteryPct != lastSentLevel) {
            lastSentLevel = batteryPct
            batteryLevelEventSink?.success(batteryPct)
        }
    }
}

File: ios/flutter_eco_mode/Sources/flutter-eco-mode/listeners/BatteryLevelListener.swift
//
//  BatteryLevelListener.swift
//  flutter_eco_mode
//
//  Created by CHOUPAULT Alexis on 27/02/2026.
//

import Flutter

class BatteryLevelListener: BatteryLevelStreamHandler, DisposableStreamListener {
    private var eventSink: PigeonEventSink<Double>?
    private var lastSentLevel: Double?
    
    // MARK: - BatteryLevelStreamHandler implementation

    override func onListen(withArguments arguments: Any?, sink: PigeonEventSink<Double>) {
        self.eventSink = sink
        
        sendUpdate()
        
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(batteryLevelChanged),
            name: UIDevice.batteryLevelDidChangeNotification,
            object: nil
        )
    }
    
    override func onCancel(withArguments arguments: Any?) {
        cleanUp()
    }
    
    // MARK: - DisposableStreamListener implementation
    
    func register(binaryMessenger: FlutterBinaryMessenger) {
        BatteryLevelStreamHandler.register(with: binaryMessenger, streamHandler: self)
    }
    
    func dispose(binaryMessenger: FlutterBinaryMessenger) {
        cleanUp()
        
        FlutterEventChannel(name: EcoModeEventChannels.batteryLevel, binaryMessenger: binaryMessenger, codec: messagesPigeonMethodCodec).setStreamHandler(nil)
    }
    
    // MARK: - Private methods

    @objc private func batteryLevelChanged() {
        sendUpdate()
    }

    private func sendUpdate() {
        let currentLevel = EcoBatteryManager.shared.getBatteryLevel()
        
        guard currentLevel != lastSentLevel else { return }
        
        lastSentLevel = currentLevel
        
        DispatchQueue.main.async { [weak self] in
            self?.eventSink?.success(currentLevel)
        }
    }
    
    private func cleanUp() {
        NotificationCenter.default.removeObserver(self)
        eventSink = nil
        lastSentLevel = nil
    }
}
